src/googleclouddebugger/__init__.py

```diff
@@ -38,42 +39,53 @@
 __version__ = version.__version__
 
 _flags = None
-_hub_client = None
+_backend_client = None
 _breakpoints_manager = None
 
 
 def _StartDebugger():
   """Configures and starts the debugger."""
-  global _hub_client
+  global _backend_client
   global _breakpoints_manager
 
   cdbg_native.InitializeModule(_flags)
   cdbg_native.LogInfo(
       f'Initializing Cloud Debugger Python agent version: {__version__}')
 
-  _hub_client = gcp_hub_client.GcpHubClient()
+  use_firebase = _flags.get('use_firebase')
+  if use_firebase:
+    _backend_client = firebase_client.FirebaseClient()
+  else:
+    _backend_client = gcp_hub_client.GcpHubClient()
+
   visibility_policy = _GetVisibilityPolicy()
 
   _breakpoints_manager = breakpoints_manager.BreakpointsManager(
-      _hub_client, visibility_policy)
+      _backend_client, visibility_policy)
 
   # Set up loggers for logpoints.
   collector.SetLogger(logging.getLogger())
 
   collector.CaptureCollector.pretty_printers.append(
       appengine_pretty_printers.PrettyPrinter)
 
-  _hub_client.on_active_breakpoints_changed = (
+  _backend_client.on_active_breakpoints_changed = (
       _breakpoints_manager.SetActiveBreakpoints)
-  _hub_client.on_idle = _breakpoints_manager.CheckBreakpointsExpiration
-  _hub_client.SetupAuth(
-      _flags.get('project_id'), _flags.get('project_number'),
-      _flags.get('service_account_json_file'))
-  _hub_client.SetupCanaryMode(
-      _flags.get('breakpoint_enable_canary'),
-      _flags.get('breakpoint_allow_canary_override'))
-  _hub_client.InitializeDebuggeeLabels(_flags)
-  _hub_client.Start()
+  _backend_client.on_idle = _breakpoints_manager.CheckBreakpointsExpiration
+
+  if use_firebase:
```

Comment:
Is it possible to move the client specific init up to where the client is instantiated above, and it's known the calls being made are supported on the instantiated type? I do see there is an ordering dependency in that  _breakpoints_manager depends on the backend client instance, and then some of the initialization of the client requires _breakpoints_manager.  So I'm not sure if the  SetupAuth and SetupCanaryMode could be moved up before it...

Reply:
Looks like they can!